
## Implement a Multilayer Perceptron for Data Classification using Tensorflow

## Aim: To build a multilayer perceptron for the  classification of  Iris Species-Setosa, Versicolor and Virginica
## Dataset : Iris


In [ ]:
import numpy as np
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
tf.compat.v1.global_variables_initializer()
import warnings
warnings.filterwarnings("ignore")

Instructions for updating:
non-resource variables are not supported in the long term


In [ ]:
def label_encoder(label):
  value =[]
  if label =="Iris-setosa":
    value = [1,0,0]
  elif label =="Iris-versicolor":
    value = [0,1,0]
  elif label =="Iris-virginica":
    value =[0,0,1]
  return value


def data_encode(file):
  X = []
  Y = []
  train_file =open("/content2/drive/MyDrive/Colab Notebooks/DEEP LEARNING/iris.train", "r")
  for line in train_file.read().strip().split("\n"):
    line = line.split(",")
    X.append([line[0], line[1],line[2],line[3]])
    Y.append(label_encoder(line[4]))
  return X, Y

def model(input, weights, bias):
  layer_1 = tf.add(tf.matmul(input, weights["hidden"]), bias["hidden"])
  layer_1 =tf.nn.relu(layer_1)
  output_layer =tf.matmul(layer_1, weights["output"])+bias["output"]
  return output_layer

X_train, Y_train = data_encode("iris.train")
X_test, Y_test = data_encode("iris.test")
learning_rate =0.01
training_epochs = 2000
display_steps = 200
n_input= 4
n_hidden = 10
n_output = 3

X = tf.compat.v1.placeholder("float", [None, n_input])
Y = tf.compat.v1.placeholder("float", [None, n_output])

weights = {
    "hidden": tf.Variable(tf.random.normal([n_input, n_hidden]), name ="Weights_hidden"),
    "output": tf.Variable(tf.random.normal([n_hidden,n_output]), name = "wieghts_output")
}

bias = {
    "hidden":tf.Variable(tf.random.normal([n_hidden]), name="bias_hidden"),
    "output": tf.Variable(tf.random.normal([n_output]), name="bias_output")
}

pred = model(X, weights, bias)

cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits =pred, labels =Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate).minimize(cost)

init = tf.global_variables_initializer()

with tf.compat.v1.Session() as sess:
  sess.run(init)
  for epoch in range(training_epochs):
    i,c = sess.run([optimizer, cost], feed_dict={X:X_train, Y:Y_train})
    if(epoch+1)% display_steps ==0:
      print("Epoch", epoch+1, "Cost: ", c)

  print("Optimization Finished")

  test_result =sess.run(pred, feed_dict={X:X_test})
  correct_pred =tf.equal(tf.argmax(test_result,1), tf.argmax(Y_train, 1))
  accuracy =tf.reduce_mean(tf.cast(correct_pred, "float"))
  ac= accuracy.eval({X:X_test, Y:Y_test})
  print("Accuracy: ", round(ac*100,2))


Instructions for updating:

Future major versions of TensorFlow will allow gradients to flow
into the labels input on backprop by default.

See `tf.nn.softmax_cross_entropy_with_logits_v2`.



Epoch 200 Cost:  0.20005403
Epoch 400 Cost:  0.15498608
Epoch 600 Cost:  0.1317955
Epoch 800 Cost:  0.117578425
Epoch 1000 Cost:  0.10798618
Epoch 1200 Cost:  0.10108706
Epoch 1400 Cost:  0.09588959
Epoch 1600 Cost:  0.09183343
Epoch 1800 Cost:  0.08857835
Epoch 2000 Cost:  0.08590703
Optimization Finished
Accuracy:  98.33


Conclusion: We can observe that as the number of epochs increased, the cost/loss has decreased considerably and the model has achieved an overall accuracy of 98%